architecture/2024-07-user-defined-types.md

```diff
@@ -0,0 +1,495 @@
+
+* **Status**: Pending/Approved
+* **Author**: `Ryan Nowak` (`@rynowak`)
+
+## Overview
+
+*This document is a high-level overview of the user-defined types feature. The scope of user-defined types is BIG. So we'll have several follow-up documents to dig into details where important.*
+
+Resources are the unit of exchange in the Radius API. All of our user-experiences like the CLI, Dashboard, and Bicep authoring are oriented around the catalog of resources. Because Radius is strongly-typed, the set of resource types and their schemas are also defined as a primary concept in Radius. In order for Radius to support each additional kind of resources, for example a PostgreSQL database, we must first have a type definition.
+
+However, the set of resource types in Radius are currently *system-defined*. Users cannot define additional resource types without making a custom build of Radius. The *user-defined-types* feature enables end-users to define their own resource types as part of their tenant of Radius. User-defined types have the same set of capabilities and enable the same features (connections, app graph, recipes, etc) as system-defined types in Radius. 
+
+Since our goal is to enable a broad community, we want to lower the barrier of entry for creating a new type. This document will define a new model for creating user-defined types based on a dynamic and declarative implementation. The declarative approach for defining a user-defined type will be supported alongside the existing support for imperative resource providers.
```

Comment:
I was wondering if there would such thing as a user-defined type becoming/graduating to be a system-defined type because it is requested by the community. Like a user-defined type being very popular or something like that.

Or are we going to have a separate repository for all user-defined types and users can register that with a command to their Radius installation?

Reply:
By graduation, I mean coming right out of the box with a fresh Radius installation.

Reply:
> I was wondering if there would such thing as a user-defined type becoming/graduating to be a system-defined type because it is requested by the community.

I think this is the right idea but the wrong terminology. 

`system-defined` -> Not modifiable by users.

Resource Group is an example of something that should probably be system-defined.

---

I think the spirit of what you're asking is absolutely a goal. We want to ship more functionality to users! If someone from the community defines a type for PostgreSQL, and it's good, we really want to put it in the box.

> By graduation, I mean coming right out of the box with a fresh Radius installation.

Yes, like that.